fix substitute problem with inccommand

erw7 opened the pull request:
fixes #12873, #12932.

Perhaps this is a more correct correction than #12947.

erw7 committed:
extmark: fix decoration ploblems with extmark

54ce101 changed the way undo entries are created when adding decorations.
This creates all sorts of problems.This change fixes the problem by
reverting to the previous behavior.

erw7 committed:
undo: fix possibility of aborting

vigoux commented on src/nvim/undo.c:
```diff
@@ -3029,15 +3029,14 @@ u_header_T *u_force_get_undo_header(buf_T *buf)
         abort();
       }
     }
+    curbuf = save_curbuf;
```
I did some testing and everything seems to work, even without this change. Is this change needed anymore ?

teto replied:
might be worth investigating how easy it would be to pass a buffer to u_savecommon

vigoux replied:
Yep, definitely, adding that to my `decurbuf` roadmap then.

erw7 replied:
In the absence of this change, the following steps can cause a crash similar to #12873.

```vim
nvim -u NORC
:vsplit file
:let bufnr = 1
:call nvim_buf_set_option(bufnr, "undofile", v:false)
:call nvim_buf_set_option(bufnr, "undolevels", -1)
:let g:mark_ns = nvim_create_namespace('myns')
:let g:mark_id = nvim_buf_set_extmark(1, g:mark_ns, 0, 0, {})
```

erw7 replied:
Since `uhp` cannot be `NULL` except when `undoleves` is -1, shouldn't the interior of the following `if` clause never be executed? Doesn't the internal execution of the `if` clause below imply that the creation or deletion of the `extmark`, without editing, creates an entry for undo/redo?

https://github.com/neovim/neovim/blob/4f8d98e583beb4c1abd5d57b9898548396633030/src/nvim/undo.c#L3024-L3041

vigoux reviewed (commented):
LGTM, just a little question in `undo.c`

erw7 commented:
@bfredl Shouldn't the following section also be `kExtmarkNoUndo`?

https://github.com/neovim/neovim/blob/4f8d98e583beb4c1abd5d57b9898548396633030/src/nvim/api/buffer.c#L1422-L1423

Unnecessary undo entries are created, as shown in the following steps.

```vim
:let g:mark_ns = nvim_create_namespace('myns')
:let g:mark_id = nvim_buf_set_extmark(1, g:mark_ns, 0, 0, {})
:undolist
number changes  when               saved
     1       1  3 seconds ago
Press ENTER or type command to continue
```

I think it's odd that just creating an extmark creates an undo entry because it says the following in `:h api`.

> Creating and deleting extmarks is not a buffer change, thus new undo states are not created for extmark changes.

bfredl commented:
@erw7 that's right. Extmark undo needs a major overhaul (by me probably, in October), to actually implement the semantics that is documented and in a no-waste way. If the existing tests pass by changing it to NoUndo then let's do that for now.

erw7 committed:
extmark: Change nvim_buf_set_extmark to not create undo data

teto commented:
@erw7 feel free to merge if it's ready

erw7 commented:
I'm not sure what the cause is, but it may or may not terminate abnormally due to a delay in input, as shown below.

```vim
nivm -u NORC
:terminal nvim -u NORC -c "set inccommand=split undolevels=-1"
" No abnormal termination.
:call chansend(b:terminal_job_id, "itest\<ESC>%s/test/f")
" Abnormal termination.
:call chansend(b:terminal_job_id, "itest\<ESC>%s/test")
:call chansend(b:terminal_job_id, "/")
:call chansend(b:terminal_job_id, "f")
```

I have modified the test based on the above findings. Based on the above findings, I changed the test to divide the input into multiple inputs and delay them by sleep. In my environment, this change causes the test to fail if this PR not applied (use ASAN build). I feel like there's room for improvement, any ideas?

If the CI is successful and there is no input from anyone to improve this, merge this.

vigoux commented:
LGTM let's merge this

erw7 commented:
I should have used `screen:expect` instead of `sleep`, as shown below.

```diff
diff --git a/test/functional/ui/inccommand_spec.lua b/test/functional/ui/inccommand_spec.lua
index 16c5477ee..e432248bc 100644
--- a/test/functional/ui/inccommand_spec.lua
+++ b/test/functional/ui/inccommand_spec.lua
@@ -2756,9 +2756,21 @@ it(":substitute doesn't crash with inccommand, if undo is empty #12932", functio
   command('set undolevels=-1')
   common_setup(screen, 'split', 'test')
   feed(':%s/test')
-  sleep(100)
+  screen:expect([[
+  {12:test}        |
+  {15:~           }|
+  {15:~           }|
+  {15:~           }|
+  :%s/test^    |
+  ]])
   feed('/')
-  sleep(100)
+  screen:expect([[
+              |
+  {15:~           }|
+  {15:~           }|
+  {15:~           }|
+  :%s/test/^   |
+  ]])
   feed('f')
   screen:expect([[
   {12:f}           |
```
